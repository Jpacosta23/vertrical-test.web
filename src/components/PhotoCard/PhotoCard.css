.photo-image-container {
  flex-basis: 20%;
}

.btn-redirect {
  background: none;
  border: none;
  cursor: pointer;
  margin: 5px 0px;
  padding: 0px;
  font-weight: bold;
  font-size: 22px;
  color: #2f81ff;
}

.photo-container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  border: 1px solid;
  border-radius: 10px;
  border-color: transparent;
  background-color: #ecfaff;
  box-shadow: 0 3px 8px 0 rgba(0, 0, 0, 0.05);
  max-width: 740px;
  margin: 30px auto;
  overflow: hidden;
  height: 250px;
}
.photo-image {
  object-fit: cover;
}

.description-container {
  width: 80%;
  display: flex;
  flex-basis: 80%;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  padding-left: 20px;
  padding-top: 10px;
}

.description-container p {
  color: #71a1a4;
}

.btn-photo {
  background-color: #2f81ff;
  border: none;
  color: white;
  padding: 16px 40px;
  border-radius: 5px;
  cursor: pointer;
  width: 250px;
}

.btn-photo:hover {
  opacity: 0.8;
}
