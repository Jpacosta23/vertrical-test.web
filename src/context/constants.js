export const GET_PHOTOS = "get-photos";
export const SET_LOADING = "set-loading";
export const GET_PHOTO = "get-photo";
export const DELETE_PHOTO = "delete-photo";
