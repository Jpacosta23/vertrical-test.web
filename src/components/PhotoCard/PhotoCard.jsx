import React from "react";
import PropTypes from "prop-types";
import "./PhotoCard.css";

const PhotoCard = ({ photo, handleRedirect = () => {} }) => {
    return (
        <div key={photo.id} className="photo-container">
            <div className="photo-image-container">
                <img
                    src={photo.photo}
                    alt={photo.short_description}
                    width="200"
                    height="250"
                    className="photo-image"
                />
            </div>
            <div className="description-container">
                <button
                    className="btn-redirect"
                    onClick={() => handleRedirect(photo)}
                >
                    {photo.title}
                </button>
                <p>{photo.short_description}</p>
                <button
                    type="button"
                    className="btn-photo"
                    onClick={() => handleRedirect(photo)}
                >
                    Go to {photo.title}
                </button>
            </div>
        </div>
    );
};

PhotoCard.propTypes = {
    photo: PropTypes.object,
    handleRedirect: PropTypes.func,
};

export default PhotoCard;
