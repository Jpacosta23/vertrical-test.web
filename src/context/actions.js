import { SET_LOADING, GET_PHOTOS, GET_PHOTO } from "./constants";
import {
    getAllPhotos,
    getPhotoById,
    getPhotosByTitle,
} from "../services/photo.service";

export async function getPhotos(dispatch) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photos = await getAllPhotos();
        dispatch({ type: GET_PHOTOS, payload: photos });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}

export async function getPhotosById(dispatch, id) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photo = await getPhotoById(id);
        dispatch({ type: GET_PHOTO, payload: photo });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}

export async function getPhotosFilteredByTitle(dispatch, title) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photos = await getPhotosByTitle(title);
        dispatch({ type: GET_PHOTOS, payload: photos });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}
