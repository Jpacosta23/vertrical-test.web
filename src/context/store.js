import React, { createContext, useReducer, useContext } from 'react';

import { GET_PHOTOS, SET_LOADING, GET_PHOTO, DELETE_PHOTO } from './constants';

const AppStateContext = createContext();
const AppDispatchContext = createContext();

const initialState = {
  photos: [],
  currentPhoto: {},
  isLoading: false
};

function AppReducer(state, action) {
  switch (action.type) {
    case GET_PHOTOS: {
      return {
        ...state,
        photos: action.payload
      };
    }
    case GET_PHOTO: {
      return {
        ...state,
        currentPhoto: action.payload
      };
    }
    case DELETE_PHOTO: {
      return {
        ...state,
        currentPhoto: action.payload
      };
    }
    case SET_LOADING: {
      return {
        ...state,
        loading: action.payload
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
}

/* eslint-disable */
function AppProvider({ children }) {
  const [state, dispatch] = useReducer(AppReducer, initialState);
  return (
    <AppStateContext.Provider value={state}>
      <AppDispatchContext.Provider value={dispatch}>{children}</AppDispatchContext.Provider>
    </AppStateContext.Provider>
  );
}

function useAppState() {
  const context = useContext(AppStateContext);
  if (context === undefined) {
    throw new Error('useAppState must be used within a AppProvider');
  }
  return context;
}

function useAppDispatch() {
  const context = useContext(AppDispatchContext);
  if (context === undefined) {
    throw new Error('useAppDispatch must be used within a AppProvider');
  }
  return context;
}

export { AppProvider, useAppState, useAppDispatch };
