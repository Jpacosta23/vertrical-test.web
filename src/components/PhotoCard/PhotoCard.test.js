import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import PhotoCard from './PhotoCard';
import renderer from 'react-test-renderer';
import { AppProvider } from '../../context/store';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate // Return an empty jest function to test whether it was called or not...I'm not depending on the results so no need to put in a return value
}));

const mockedProps = {
  id: 1,
  title: 'profile',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
  photo:
    'https://i.picsum.photos/id/1027/2848/4272.jpg?hmac=EAR-f6uEqI1iZJjB6-NzoZTnmaX0oI0th3z8Y78UpKM',
  short_description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.'
};

afterEach(cleanup);
describe('Not Found component test', () => {
  it('renders without crashing', () => {
    const props = {
      photo: mockedProps
    };
    const div = document.createElement('div');
    ReactDOM.render(
      <AppProvider>
        <MemoryRouter>
          <PhotoCard photo={props.photo} />
        </MemoryRouter>
      </AppProvider>,
      div
    );
  });
  it('matches snapshot', () => {
    const props = {
      photo: mockedProps
    };
    const tree = renderer
      .create(
        <AppProvider>
          <MemoryRouter>
            <PhotoCard photo={props.photo} />
          </MemoryRouter>
        </AppProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
