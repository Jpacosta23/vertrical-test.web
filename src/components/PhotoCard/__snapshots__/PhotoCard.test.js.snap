// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not Found component test matches snapshot 1`] = `
<div
  className="photo-container"
>
  <div
    className="photo-image-container"
  >
    <img
      alt="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat."
      className="photo-image"
      height="250"
      src="https://i.picsum.photos/id/1027/2848/4272.jpg?hmac=EAR-f6uEqI1iZJjB6-NzoZTnmaX0oI0th3z8Y78UpKM"
      width="200"
    />
  </div>
  <div
    className="description-container"
  >
    <button
      className="btn-redirect"
      onClick={[Function]}
    >
      profile
    </button>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    </p>
    <button
      className="btn-photo"
      onClick={[Function]}
      type="button"
    >
      Go to 
      profile
    </button>
  </div>
</div>
`;
